{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b67c13b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import sys, getopt\n",
    "import xml.etree.ElementTree as et \n",
    "import json\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import numpy as np\n",
    "#from pyodide.http import pyfetch\n",
    "#from js import XMLHttpRequest\n",
    "# print(input_json)\n",
    "with open(\"./inv.json\", \"r\") as fh:\n",
    "    input_json  = json.load(fh)\n",
    "\n",
    "input_json = str(input_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "55096b59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start\n",
      "Time used for conversion is: 0.32 seconds\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "start = time.time()\n",
    "print(\"start\")\n",
    "def append_table_protocol(element, json_rows):\n",
    "\n",
    "    for parameter in element:\n",
    "\n",
    "        p=parameter.get('parameterName') if parameter.get('parameterName') is not None else None\n",
    "#         #  print(p)\n",
    "        term_name = p.get('annotationValue') if p.get('annotationValue') is not None else None\n",
    "        #  print('name is: '+ term_name )\n",
    "        term_source = p.get('termSource') if p.get('termSource')  is not None else None\n",
    "        #  print( term_source)\n",
    "        term_accession = p.get('termAccession') if p.get('termAccession')  is not None else None\n",
    "        #  print( term_accession)                          \n",
    "        index_value= p.get('comments')[0].get(\"value\")\n",
    "        json_rows.append( [file_id,\n",
    "                           arc_id, \n",
    "                           subsection_id,\n",
    "                           input_id, \n",
    "                           input_name,\n",
    "\n",
    "                           output_id, \n",
    "                           output_name,\n",
    "                           derived_from,\n",
    "                           term_type,\n",
    "                           term_name, \n",
    "                           term_source, \n",
    "                           term_accession, \n",
    "                           index_value, \n",
    "                           value])\n",
    "#         print('index value is: '+ term_value)\n",
    "        \n",
    "\n",
    "\n",
    "def append_table_assays(element, json_rows, option= 1):\n",
    "    \n",
    "    for parameter in element:\n",
    "\n",
    "        c= parameter.get(\"category\")\n",
    "        p=c.get('parameterName') if c.get('parameterName') is not None else None\n",
    "#         print(p)\n",
    "        value = parameter.get('value') if parameter.get('value') is not None else None\n",
    "        term_name = p.get('annotationValue') if p.get('annotationValue') is not None else None\n",
    "#         print('name is: '+ term_name )\n",
    "        term_source = p.get('termSource') if (p.get('termSource') ) is not None else None\n",
    "#         print( term_source)\n",
    "        term_accession = p.get('termAccession') if (p.get('termAccession') ) is not None else None\n",
    "#         print( term_accession)\n",
    "        term_type = \"parameter\"\n",
    "        index_value= p.get('comments')[0].get(\"value\")\n",
    "        \n",
    "        json_rows.append( [file_id, \n",
    "                           arc_id, \n",
    "                           subsection_id,\n",
    "                           input_id, \n",
    "                           input_name,\n",
    "\n",
    "                           output_id, \n",
    "                           output_name,\n",
    "                           derived_from,\n",
    "                           term_type,\n",
    "                           term_name, \n",
    "                           term_source, \n",
    "                           term_accession, \n",
    "                           index_value, \n",
    "                           value])\n",
    "#         print('index value is: '+ term_value)\n",
    "        \n",
    "def append_c_input(element, json_rows):\n",
    "    \n",
    "    for parameter in element:\n",
    "\n",
    "        b = parameter.get(\"category\")\n",
    "        factor = b.get(\"characteristicType\") if b.get(\"characteristicType\") is not None else None\n",
    "#         print(p)\n",
    "        value = parameter.get('value') if parameter.get('value') is not None else None\n",
    "        term_name = factor.get('annotationValue') if factor.get('annotationValue') is not None else None\n",
    "#         print('name is: '+ term_name )\n",
    "        term_source = factor.get('termSource') if (factor.get('termSource') ) is not None else None\n",
    "#         print( term_source)\n",
    "        term_accession = factor.get('termAccession') if (factor.get('termAccession') ) is not None else None\n",
    "#         print( term_accession)\n",
    "        term_type = \"parameter\"\n",
    "        index_value= factor.get('comments')[0].get(\"value\")\n",
    "        json_rows.append( [file_id, \n",
    "                           arc_id, \n",
    "                           subsection_id,\n",
    "                           input_id, \n",
    "                           input_name,\n",
    "\n",
    "                           output_id, \n",
    "                           output_name,\n",
    "                           derived_from,\n",
    "                           term_type,\n",
    "                           term_name, \n",
    "                           term_source, \n",
    "                           term_accession, \n",
    "                           index_value, \n",
    "                           value])\n",
    "#         print('index value is: '+ term_value)\n",
    "\n",
    "def append_input(element, json_rows):\n",
    "    \n",
    "    for parameter in element:\n",
    "\n",
    "        b = parameter.get(\"category\")\n",
    "        factor = b.get(\"factorType\") if b.get(\"factorType\") is not None else None\n",
    "#         print(p)\n",
    "\n",
    "        value = parameter.get('value') if parameter.get('value') is not None else None\n",
    "        term_name = factor.get('annotationValue') if factor.get('annotationValue') is not None else None\n",
    "#         print('name is: '+ term_name )\n",
    "        term_source = factor.get('termSource') if (factor.get('termSource') ) is not None else None\n",
    "#         print( term_source)\n",
    "        term_accession = factor.get('termAccession') if (factor.get('termAccession') ) is not None else None\n",
    "#         print( term_accession)\n",
    "        term_type = \"factor\"\n",
    "        index_value= factor.get('comments')[0].get(\"value\")\n",
    "        json_rows.append( [file_id, \n",
    "                           arc_id, \n",
    "                           subsection_id,\n",
    "                           input_id, \n",
    "                           input_name,\n",
    "\n",
    "                           output_id, \n",
    "                           output_name,\n",
    "                           derived_from,\n",
    "                           term_type,\n",
    "                           term_name, \n",
    "                           term_source, \n",
    "                           term_accession, \n",
    "                           index_value, \n",
    "                           value])\n",
    "#         print('index value is: '+ term_value)\n",
    "\n",
    "def append_output(element, json_rows):\n",
    "    \n",
    "    for parameter in element:\n",
    "        \n",
    "        b = parameter.get(\"category\")\n",
    "        factor = b.get(\"factorType\") if b.get(\"factorType\") is not None else None\n",
    "#         print(p)\n",
    "\n",
    "        value = parameter.get('value') if parameter.get('value') is not None else None\n",
    "        term_name = factor.get('annotationValue') if factor.get('annotationValue') is not None else None\n",
    "#         print('name is: '+ term_name )\n",
    "        term_source = factor.get('termSource') if (factor.get('termSource') ) is not None else None\n",
    "#         print( term_source)\n",
    "        term_accession = factor.get('termAccession') if (factor.get('termAccession') ) is not None else None\n",
    "#         print( term_accession)\n",
    "        term_type = 'factor'\n",
    "        index_value= factor.get('comments')[0].get(\"value\")\n",
    "        json_rows.append( [file_id, \n",
    "                           arc_id, \n",
    "                           subsection_id,\n",
    "                           input_id, \n",
    "                           input_name,\n",
    "\n",
    "                           output_id, \n",
    "                           output_name,\n",
    "                           derived_from,\n",
    "                           term_type,\n",
    "                           term_name, \n",
    "                           term_source, \n",
    "                           term_accession, \n",
    "                           index_value, \n",
    "                           value])\n",
    "#         print('index value is: '+ term_value)\n",
    "\n",
    "# todo factorValues, derived from\n",
    "\n",
    "def init_none():\n",
    "    global input_id \n",
    "    global input_name \n",
    "    global output_id\n",
    "    global output_name \n",
    "    global derived_from \n",
    "    global value \n",
    "    global index_value\n",
    "    global term_type\n",
    "    global term_name\n",
    "    global term_source \n",
    "    global term_accession \n",
    "    input_id = None\n",
    "    input_name = None\n",
    "    output_id = None\n",
    "    output_name = None\n",
    "    derived_from = None\n",
    "    value = None\n",
    "    index_value = None\n",
    "    term_type = None\n",
    "    term_name = None\n",
    "    term_source = None\n",
    "    term_accession = None\n",
    "    \n",
    "def write_sample_xml(xml_input):\n",
    "    input_ = xml_input[xml_input[\"subsection_id\"].str.match(r'1SPL01_plants_')==True ]\n",
    "    sample= input_\n",
    "    input_list = sample[sample[\"input_name\"].notna()]\n",
    "    output_list = sample[sample[\"output_name\"].notna()]\n",
    "\n",
    "    a = et.Element('SAMPLE_SET')\n",
    "    for index1,input_name in enumerate(input_list[\"input_name\"].unique()):\n",
    "        one_sample = sample[sample['input_name'] == input_name]\n",
    "        for index2, tags in enumerate(one_sample):\n",
    "            b = et.SubElement(a, 'SAMPLE')\n",
    "\n",
    "            b.attrib = {'alias': input_name , 'center_name':''}\n",
    "            c = et.SubElement(b, 'TITLE')\n",
    "            c.text = one_sample[one_sample['term_name']== 'Organism']['value'].to_string(header=False, index=False)\n",
    "\n",
    "            d = et.SubElement(b, 'SAMPLE_NAME')\n",
    "            #d.text = input_name\n",
    "\n",
    "            e = et.SubElement(d, 'TAXON_ID')\n",
    "            e.text = '110664'\n",
    "            f = et.SubElement(d, 'SCIENTIFIC_NAME')\n",
    "            f.text = c.text = one_sample[one_sample['term_name']== 'Organism']['value'].to_string(header=False, index=False)\n",
    "\n",
    "            g = et.SubElement(d, 'COMMON_NAME')\n",
    "            g.text = input_name\n",
    "\n",
    "            h = et.SubElement(b, 'SAMPLE_ATTRIBUTES')\n",
    "            for index3, row in one_sample.iterrows():\n",
    "                m = et.SubElement(h,'SAMPLE_ATTRIBUTE' )\n",
    "                j = et.SubElement(m, 'TAG')\n",
    "                j.text = str(row['term_name'])\n",
    "\n",
    "                k = et.SubElement(m, 'VALUE')\n",
    "                k.text = str(row['value']) \n",
    "\n",
    "\n",
    "    tree = et.ElementTree(a)\n",
    "    et.indent(tree, space=\" \", level=0)\n",
    "    #et.dump(a)\n",
    "    tree.write(\"SAMPLE.xml\", encoding=\"utf-8\")\n",
    "    return tree\n",
    "        \n",
    "\n",
    "def convert_json(arc_json_file, mapping_file):\n",
    "    with open(arc_json_file, \"r\") as read_file:\n",
    "#     print(\"Converting JSON encoded data into Python dictionary\")\n",
    "        arc_json = json.load(read_file)\n",
    "    \n",
    "    \n",
    "\n",
    "    global xml_output\n",
    "    global xmloutput\n",
    "    global file_id\n",
    "    global arc_id\n",
    "    global subsection_id\n",
    "    global input_id \n",
    "    global input_name \n",
    "    global output_id\n",
    "    global output_name \n",
    "    global derived_from \n",
    "    global value \n",
    "    global index_value\n",
    "    global term_type\n",
    "    global term_name\n",
    "    global term_source \n",
    "    global term_accessio\n",
    "    json_table= []\n",
    "    json_cols = [\"file_id\", \n",
    "                 \"arc_id\" , \n",
    "                 \"subsection_id\", \n",
    "                 \"input_id\", \n",
    "                 \"input_name\", \n",
    "\n",
    "                 \"output_id\", \n",
    "                 \"output_name\",\n",
    "                 \"derived_from\",\n",
    "                 \"term_type\",\n",
    "                 \"term_name\", \n",
    "                 \"term_source\", \n",
    "                 \"term_accession\", \n",
    "                 \"index_value\", \n",
    "                 \"value\", ]\n",
    "    json_rows = []\n",
    "    init_none()\n",
    "    file_id = arc_json.get(\"identifier\")\n",
    "    section = \"studies\"\n",
    "    studies = arc_json.get(\"studies\")\n",
    "    arc_id = studies[0].get(\"identifier\")\n",
    "    processSequences = arc_json.get(\"studies\")[0].get('assays')[0].get(\"processSequence\")\n",
    "\n",
    "    for process_seq in processSequences:\n",
    "\n",
    "        subsection_id = process_seq.get(\"name\")\n",
    "        #  print(subsection_id)\n",
    "\n",
    "        parameter_values = process_seq.get(\"parameterValues\")\n",
    "        append_table_assays(parameter_values, json_rows)\n",
    "        init_none() \n",
    "        input_section = process_seq.get(\"inputs\")\n",
    "        if input_section != None:\n",
    "    #     print(input_section)\n",
    "            for i,input_ in enumerate(input_section):\n",
    "                input_name = input_.get(\"name\")\n",
    "                input_id = i\n",
    "\n",
    "                derived_from = input_.get(\"derivesFrom\")[0].get(\"name\") if input_.get(\"derivesFrom\") is not None else None\n",
    "                #  print(input_name + \" derivesFrom is \", derived_from)\n",
    "                characteristics = input_.get(\"characteristics\") \n",
    "                if characteristics != None:\n",
    "                    append_c_input(characteristics,json_rows)\n",
    "                factorvalues = input_.get(\"factorValues\") \n",
    "                if factorvalues != None:\n",
    "                        append_input(factorvalues,json_rows)\n",
    "\n",
    "\n",
    "\n",
    "            \n",
    "        output_section = process_seq.get(\"outputs\")\n",
    "\n",
    "    #     print(output_section)\n",
    "        for i, output_ in enumerate(output_section):\n",
    "            output_name = output_.get(\"name\")\n",
    "            output_id = i\n",
    "            derived_from = output_.get(\"derivesFrom\")[0].get(\"name\") if output_.get(\"derivesFrom\") is not None else None\n",
    "            #  print(output_name + \" and derived from is \", derived_from)\n",
    "\n",
    "            term_type = output_.get('type') if output_.get('type') is not None else None\n",
    "\n",
    "            factorvalues = output_.get(\"factorValues\") \n",
    "            if factorvalues != None:\n",
    "                append_output(factorvalues,json_rows)\n",
    "            else:\n",
    "                json_rows.append( [file_id, \n",
    "                               arc_id, \n",
    "                               subsection_id,\n",
    "                               input_id, \n",
    "                               input_name,\n",
    "\n",
    "                               output_id, \n",
    "                               output_name,\n",
    "                               derived_from,\n",
    "                                   term_type,\n",
    "                               term_name, \n",
    "                               term_source, \n",
    "                               term_accession, \n",
    "                               index_value, \n",
    "                               value])\n",
    "\n",
    "\n",
    "\n",
    "    json_table = pd.DataFrame(json_rows, columns= json_cols)\n",
    "    mapping = pd.read_csv(mapping_file)\n",
    "    term_id = mapping.iloc[:,2]\n",
    "    mapping.iloc[:,3] = term_id.map( lambda x: \"http://purl.obolibrary.org/obo/{}\".format( str(x).replace(\":\", \"_\") )  )\n",
    "\n",
    "    \n",
    "    json_table[\"ena_id\"] = json_table[\"term_name\"]\n",
    "    a = json_table.loc[:, \"ena_id\"]\n",
    "    # a = json_table.loc[:, \"term_accession\"]\n",
    "    term_mapping = mapping.iloc[:,3]\n",
    "    for i, term in enumerate(term_mapping):\n",
    "         cond = json_table[\"term_accession\"] == term\n",
    "    #     print(term)\n",
    "    #     print(filter_)\n",
    "         a.mask(cond , mapping.loc[:, \"Field Name\"][i], inplace= True)\n",
    "    json_table.to_csv('json_table_new.csv')\n",
    "    \n",
    "    \n",
    "    a = et.Element('PROJECT_SET')\n",
    "    for i,project in enumerate(arc_json.get(\"studies\")):\n",
    "        b = et.SubElement(a, 'PROJECT')\n",
    "\n",
    "        b.attrib = {'alias': project.get('identifier')}\n",
    "        c = et.SubElement(b, 'TITLE')\n",
    "        c.text = project.get('title')\n",
    "        d = et.SubElement(b, 'DESCRIPTION')\n",
    "        d.text = str(arc_json.get(\"publications\"))\n",
    "        f = et.SubElement(b, \"SUBMISSION_PROJECT\")\n",
    "        g = et.SubElement(f, \"SEQUENCING_PROJECT\")\n",
    "    tree = et.ElementTree(a)\n",
    "    et.indent(tree, space=\" \", level=0)\n",
    "\n",
    "    tree.write(\"STUDY.xml\", encoding=\"utf-8\")\n",
    "\n",
    "    # sample xml\n",
    "    xml_input = pd.read_csv(\"json_table_new.csv\")\n",
    "    xml_output = et.tostring(write_sample_xml(xml_input).getroot(), encoding='utf8', method='xml')\n",
    "    xmloutput = xml_output.decode(\"utf-8\")\n",
    "    with open(\"SAMPLE.XML\", \"wb\") as fh:\n",
    "        fh.write(xml_output )\n",
    "    \n",
    "#req1 = XMLHttpRequest.new()\n",
    "#req1.open(\"GET\", \"./inv.json\", False)\n",
    "#req1.send(None)\n",
    "#arc_json_file = str(req1.response)\n",
    "# with open(\"inv.json\", \"w\") as fh:\n",
    "#       fh.write(input_json)\n",
    "\n",
    "# req2 = XMLHttpRequest.new()\n",
    "# req2.open(\"GET\", \"./mapping_ERC000037.csv\", False)\n",
    "# req2.send(None)\n",
    "# mapping_file = str(req2.response)\n",
    "# with open(\"mapping_ERC000037.csv\", \"w\") as fh:\n",
    "#       fh.write(mapping_file)\n",
    "\n",
    "\n",
    "#req3 = XMLHttpRequest.new()\n",
    "#req3.open(\"GET\", \"./json_table_new.csv\", False)\n",
    "#req3.send(None)\n",
    "#json_table_new = str(req3.response)\n",
    "#with open(\"json_table_new.csv\", \"w\") as fh:\n",
    "#      fh.write(json_table_new)\n",
    "\n",
    "convert_json(\"inv.json\", \"mapping_ERC000037.csv\")\n",
    "end = time.time()\n",
    "used_time = end - start\n",
    "print(\"Time used for conversion is: {} seconds\".format(round(used_time, 2)) )\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "edbc42d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "arc_json_file = \"inv.json\"\n",
    "mapping_file=\"mapping_ERC000037.csv\"\n",
    "with open(arc_json_file, \"r\") as read_file:\n",
    "#     print(\"Converting JSON encoded data into Python dictionary\")\n",
    "    arc_json = json.load(read_file)\n",
    "\n",
    "\n",
    "\n",
    "global xml_output\n",
    "global xmloutput\n",
    "global file_id\n",
    "global arc_id\n",
    "global subsection_id\n",
    "global input_id \n",
    "global input_name \n",
    "global output_id\n",
    "global output_name \n",
    "global derived_from \n",
    "global value \n",
    "global index_value\n",
    "global term_type\n",
    "global term_name\n",
    "global term_source \n",
    "global term_accessio\n",
    "json_table= []\n",
    "json_cols = [\"file_id\", \n",
    "             \"arc_id\" , \n",
    "             \"subsection_id\", \n",
    "             \"input_id\", \n",
    "             \"input_name\", \n",
    "\n",
    "             \"output_id\", \n",
    "             \"output_name\",\n",
    "             \"derived_from\",\n",
    "             \"term_type\",\n",
    "             \"term_name\", \n",
    "             \"term_source\", \n",
    "             \"term_accession\", \n",
    "             \"index_value\", \n",
    "             \"value\", ]\n",
    "json_rows = []\n",
    "init_none()\n",
    "file_id = arc_json.get(\"identifier\")\n",
    "section = \"studies\"\n",
    "studies = arc_json.get(\"studies\")\n",
    "arc_id = studies[0].get(\"identifier\")\n",
    "processSequences = arc_json.get(\"studies\")[0].get('assays')[0].get(\"processSequence\")\n",
    "\n",
    "for process_seq in processSequences:\n",
    "\n",
    "    subsection_id = process_seq.get(\"name\")\n",
    "    #  print(subsection_id)\n",
    "\n",
    "    parameter_values = process_seq.get(\"parameterValues\")\n",
    "    append_table_assays(parameter_values, json_rows)\n",
    "    init_none() \n",
    "    input_section = process_seq.get(\"inputs\")\n",
    "    if input_section != None:\n",
    "#     print(input_section)\n",
    "        for i,input_ in enumerate(input_section):\n",
    "            input_name = input_.get(\"name\")\n",
    "            input_id = i\n",
    "\n",
    "            derived_from = input_.get(\"derivesFrom\")[0].get(\"name\") if input_.get(\"derivesFrom\") is not None else None\n",
    "            #  print(input_name + \" derivesFrom is \", derived_from)\n",
    "            characteristics = input_.get(\"characteristics\") \n",
    "            if characteristics != None:\n",
    "                append_c_input(characteristics,json_rows)\n",
    "            factorvalues = input_.get(\"factorValues\") \n",
    "            if factorvalues != None:\n",
    "                    append_input(factorvalues,json_rows)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    output_section = process_seq.get(\"outputs\")\n",
    "\n",
    "#     print(output_section)\n",
    "    for i, output_ in enumerate(output_section):\n",
    "        output_name = output_.get(\"name\")\n",
    "        output_id = i\n",
    "        derived_from = output_.get(\"derivesFrom\")[0].get(\"name\") if output_.get(\"derivesFrom\") is not None else None\n",
    "        #  print(output_name + \" and derived from is \", derived_from)\n",
    "\n",
    "        term_type = output_.get('type') if output_.get('type') is not None else None\n",
    "\n",
    "        factorvalues = output_.get(\"factorValues\") \n",
    "        if factorvalues != None:\n",
    "            append_output(factorvalues,json_rows)\n",
    "        else:\n",
    "            json_rows.append( [file_id, \n",
    "                           arc_id, \n",
    "                           subsection_id,\n",
    "                           input_id, \n",
    "                           input_name,\n",
    "\n",
    "                           output_id, \n",
    "                           output_name,\n",
    "                           derived_from,\n",
    "                               term_type,\n",
    "                           term_name, \n",
    "                           term_source, \n",
    "                           term_accession, \n",
    "                           index_value, \n",
    "                           value])\n",
    "\n",
    "\n",
    "\n",
    "json_table = pd.DataFrame(json_rows, columns= json_cols)\n",
    "mapping = pd.read_csv(mapping_file)\n",
    "term_id = mapping.iloc[:,2]\n",
    "mapping.iloc[:,3] = term_id.map( lambda x: \"http://purl.obolibrary.org/obo/{}\".format( str(x).replace(\":\", \"_\") )  )\n",
    "\n",
    "json_table[\"in_ena\"] = False\n",
    "b = json_table.loc[:, \"in_ena\"]\n",
    "json_table[\"ena_id\"] = json_table[\"term_name\"]\n",
    "\n",
    "a = json_table.loc[:, \"ena_id\"]\n",
    "\n",
    "# a = json_table.loc[:, \"term_accession\"]\n",
    "term_mapping = mapping.iloc[:,3]\n",
    "for i, term in enumerate(term_mapping):\n",
    "    cond = json_table[\"term_accession\"] == term\n",
    "#     print(term)\n",
    "#     print(filter_)\n",
    "    a.mask(cond , mapping.loc[:, \"Field Name\"][i], inplace= True)\n",
    "    b.mask(cond , mapping.loc[:, \"Requirement\"][i], inplace= True)\n",
    "json_table.to_csv('json_table_new.csv')\n",
    "\n",
    "\n",
    "a = et.Element('PROJECT_SET')\n",
    "for i,project in enumerate(arc_json.get(\"studies\")):\n",
    "    b = et.SubElement(a, 'PROJECT')\n",
    "\n",
    "    b.attrib = {'alias': project.get('identifier')}\n",
    "    c = et.SubElement(b, 'TITLE')\n",
    "    c.text = project.get('title')\n",
    "    d = et.SubElement(b, 'DESCRIPTION')\n",
    "    d.text = str(arc_json.get(\"publications\"))\n",
    "    f = et.SubElement(b, \"SUBMISSION_PROJECT\")\n",
    "    g = et.SubElement(f, \"SEQUENCING_PROJECT\")\n",
    "tree = et.ElementTree(a)\n",
    "et.indent(tree, space=\" \", level=0)\n",
    "\n",
    "tree.write(\"STUDY.xml\", encoding=\"utf-8\")\n",
    "\n",
    "# sample xml\n",
    "xml_input = pd.read_csv(\"json_table_new.csv\")\n",
    "xml_output = et.tostring(write_sample_xml(xml_input).getroot(), encoding='utf8', method='xml')\n",
    "xmloutput = xml_output.decode(\"utf-8\")\n",
    "with open(\"SAMPLE.XML\", \"wb\") as fh:\n",
    "    fh.write(xml_output )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "abfc638c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         False\n",
       "1         False\n",
       "2      optional\n",
       "3         False\n",
       "4      optional\n",
       "         ...   \n",
       "304       False\n",
       "305       False\n",
       "306       False\n",
       "307       False\n",
       "308       False\n",
       "Name: in_ena, Length: 309, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json_table[\"in_ena\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "5893431d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>arc_id</th>\n",
       "      <th>subsection_id</th>\n",
       "      <th>input_id</th>\n",
       "      <th>input_name</th>\n",
       "      <th>output_id</th>\n",
       "      <th>output_name</th>\n",
       "      <th>derived_from</th>\n",
       "      <th>term_type</th>\n",
       "      <th>term_name</th>\n",
       "      <th>term_source</th>\n",
       "      <th>index_value</th>\n",
       "      <th>value</th>\n",
       "      <th>in_ena</th>\n",
       "      <th>ena_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>term_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000009</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000011</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000030</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000031</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000065</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000071</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000072</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000073</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000074</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000075</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/NFDI4PSO_0000076</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/PECO_0007147</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/PECO_0007383</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://purl.obolibrary.org/obo/TO_1000012</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 file_id  arc_id  \\\n",
       "term_accession                                                     \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000009        1       1   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000011        1       1   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000030        6       6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000031        6       6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000065        6       6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000071        6       6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000072        6       6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000073        6       6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000074        6       6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000075        6       6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000076        6       6   \n",
       "http://purl.obolibrary.org/obo/PECO_0007147            6       6   \n",
       "http://purl.obolibrary.org/obo/PECO_0007383            6       6   \n",
       "http://purl.obolibrary.org/obo/TO_1000012              6       6   \n",
       "\n",
       "                                                 subsection_id  input_id  \\\n",
       "term_accession                                                             \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000009              1         0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000011              1         0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000030              6         6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000031              6         6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000065              6         6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000071              6         6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000072              6         6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000073              6         6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000074              6         6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000075              6         6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000076              6         6   \n",
       "http://purl.obolibrary.org/obo/PECO_0007147                  6         6   \n",
       "http://purl.obolibrary.org/obo/PECO_0007383                  6         6   \n",
       "http://purl.obolibrary.org/obo/TO_1000012                    6         6   \n",
       "\n",
       "                                                 input_name  output_id  \\\n",
       "term_accession                                                           \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000009           0          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000011           0          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000030           6          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000031           6          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000065           6          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000071           6          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000072           6          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000073           6          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000074           6          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000075           6          0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000076           6          0   \n",
       "http://purl.obolibrary.org/obo/PECO_0007147               6          0   \n",
       "http://purl.obolibrary.org/obo/PECO_0007383               6          0   \n",
       "http://purl.obolibrary.org/obo/TO_1000012                 6          0   \n",
       "\n",
       "                                                 output_name  derived_from  \\\n",
       "term_accession                                                               \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000009            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000011            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000030            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000031            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000065            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000071            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000072            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000073            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000074            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000075            0             0   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000076            0             0   \n",
       "http://purl.obolibrary.org/obo/PECO_0007147                0             0   \n",
       "http://purl.obolibrary.org/obo/PECO_0007383                0             0   \n",
       "http://purl.obolibrary.org/obo/TO_1000012                  0             0   \n",
       "\n",
       "                                                 term_type  term_name  \\\n",
       "term_accession                                                          \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000009          1          1   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000011          1          1   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000030          6          6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000031          6          6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000065          6          6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000071          6          6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000072          6          6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000073          6          6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000074          6          6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000075          6          6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000076          6          6   \n",
       "http://purl.obolibrary.org/obo/PECO_0007147              6          6   \n",
       "http://purl.obolibrary.org/obo/PECO_0007383              6          6   \n",
       "http://purl.obolibrary.org/obo/TO_1000012                6          6   \n",
       "\n",
       "                                                 term_source  index_value  \\\n",
       "term_accession                                                              \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000009            1            1   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000011            1            1   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000030            6            6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000031            6            6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000065            6            6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000071            6            6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000072            6            6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000073            6            6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000074            6            6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000075            6            6   \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000076            6            6   \n",
       "http://purl.obolibrary.org/obo/PECO_0007147                6            6   \n",
       "http://purl.obolibrary.org/obo/PECO_0007383                6            6   \n",
       "http://purl.obolibrary.org/obo/TO_1000012                  6            6   \n",
       "\n",
       "                                                 value  in_ena  ena_id  \n",
       "term_accession                                                          \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000009      0       1       1  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000011      0       1       1  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000030      6       6       6  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000031      0       6       6  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000065      0       6       6  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000071      0       6       6  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000072      0       6       6  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000073      0       6       6  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000074      0       6       6  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000075      0       6       6  \n",
       "http://purl.obolibrary.org/obo/NFDI4PSO_0000076      0       6       6  \n",
       "http://purl.obolibrary.org/obo/PECO_0007147          6       6       6  \n",
       "http://purl.obolibrary.org/obo/PECO_0007383          6       6       6  \n",
       "http://purl.obolibrary.org/obo/TO_1000012            0       6       6  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json_table[json_table[\"in_ena\"]!=False].groupby(['term_accession']).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "24b9648b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['http://purl.obolibrary.org/obo/NFDI4PSO_0000065',\n",
       "  'http://purl.obolibrary.org/obo/NFDI4PSO_0000075',\n",
       "  'http://purl.obolibrary.org/obo/PECO_0007147'],\n",
       " ['Isolate', 'Sample Collection Date', 'plant growth medium exposure'],\n",
       " ['http://purl.obolibrary.org/obo/NFDI4PSO_0000030',\n",
       "  'http://purl.obolibrary.org/obo/NFDI4PSO_0000071',\n",
       "  'http://purl.obolibrary.org/obo/PECO_0007383'],\n",
       " ['Organism', 'Plant disease', 'watering exposure'],\n",
       " ['http://purl.obolibrary.org/obo/NFDI4PSO_0000009',\n",
       "  'http://purl.obolibrary.org/obo/NFDI4PSO_0000011',\n",
       "  'http://purl.obolibrary.org/obo/NFDI4PSO_0000031',\n",
       "  'http://purl.obolibrary.org/obo/NFDI4PSO_0000072',\n",
       "  'http://purl.obolibrary.org/obo/NFDI4PSO_0000073',\n",
       "  'http://purl.obolibrary.org/obo/NFDI4PSO_0000074',\n",
       "  'http://purl.obolibrary.org/obo/NFDI4PSO_0000076',\n",
       "  'http://purl.obolibrary.org/obo/TO_1000012'],\n",
       " ['Genotype',\n",
       "  'Geographic Area',\n",
       "  'Phenotype',\n",
       "  'Plant disease stage',\n",
       "  'Sample Collected By',\n",
       "  'Sample Collection Method',\n",
       "  'Sample storage',\n",
       "  'whole plant size']]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[list(json_table[json_table[\"in_ena\"]==\"mandatory\"].groupby(['term_accession']).count().index),\n",
    "list(json_table[json_table[\"in_ena\"]==\"mandatory\"].groupby(['term_name']).count().index),\n",
    "list(json_table[json_table[\"in_ena\"]==\"recommended\"].groupby(['term_accession']).count().index),\n",
    "list(json_table[json_table[\"in_ena\"]==\"recommended\"].groupby(['term_name']).count().index),\n",
    "list(json_table[json_table[\"in_ena\"]==\"optional\"].groupby(['term_accession']).count().index),\n",
    "list(json_table[json_table[\"in_ena\"]==\"optional\"].groupby(['term_name']).count().index),\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "a4d0c1f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([ 37,  72, 107, 142, 177, 212]),)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.where(json_table[\"term_accession\"] == 'http://purl.obolibrary.org/obo/NFDI4PSO_0000076')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "f479d82c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a= json_table[\"term_accession\"] == 'http://purl.obolibrary.org/obo/NFDI4PSO_0000076'\n",
    "a[37]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3e5a2c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping[\"in_metadata\"] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "18ab6b22",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "beb8b3e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_164/3233280867.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  [mapping[np.invert(isin)][mapping[\"Requirement\"]==\"mandatory\"]['id'],\n",
      "/tmp/ipykernel_164/3233280867.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  mapping[np.invert(isin)][mapping[\"Requirement\"]==\"mandatory\"]['Field Name'],\n",
      "/tmp/ipykernel_164/3233280867.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  mapping[np.invert(isin)][mapping[\"Requirement\"]==\"recommended\"]['id'],\n",
      "/tmp/ipykernel_164/3233280867.py:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  mapping[np.invert(isin)][mapping[\"Requirement\"]==\"recommended\"]['Field Name'],\n",
      "/tmp/ipykernel_164/3233280867.py:6: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  mapping[np.invert(isin)][mapping[\"Requirement\"]==\"optional\"]['id'],\n",
      "/tmp/ipykernel_164/3233280867.py:7: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  mapping[np.invert(isin)][mapping[\"Requirement\"]==\"optional\"]['Field Name'],\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[9     NFDI4PSO:0002003\n",
       " 10    NFDI4PSO:0002004\n",
       " 27          PO:0009011\n",
       " 28          PO:0009012\n",
       " Name: id, dtype: object,\n",
       " 9      geographic location (latitude)\n",
       " 10    geographic location (longitude)\n",
       " 27                    plant structure\n",
       " 28          plant developmental stage\n",
       " Name: Field Name, dtype: object,\n",
       " 0     NaN\n",
       " 1     NaN\n",
       " 3     NaN\n",
       " 17    NaN\n",
       " 21    NaN\n",
       " 32    NaN\n",
       " 38    NaN\n",
       " 43    NaN\n",
       " 44    NaN\n",
       " 55    NaN\n",
       " 64    NaN\n",
       " 66    NaN\n",
       " 67    NaN\n",
       " 68    NaN\n",
       " 69    NaN\n",
       " 70    NaN\n",
       " 71    NaN\n",
       " 72    NaN\n",
       " 73    NaN\n",
       " 74    NaN\n",
       " 75    NaN\n",
       " 84    NaN\n",
       " 94    NaN\n",
       " 95    NaN\n",
       " Name: id, dtype: object,\n",
       " 0                                 ploidy\n",
       " 1                    number of replicons\n",
       " 3                         estimated size\n",
       " 17           source material identifiers\n",
       " 21                           propagation\n",
       " 32                 sample disease status\n",
       " 38                       growth facility\n",
       " 43      subspecific genetic lineage rank\n",
       " 44      subspecific genetic lineage name\n",
       " 55     soil_taxonomic/FAO classification\n",
       " 64                               soil pH\n",
       " 66                    rooting conditions\n",
       " 67                culture rooting medium\n",
       " 68         rooting medium macronutrients\n",
       " 69         rooting medium micronutrients\n",
       " 70    rooting medium organic supplements\n",
       " 71                 rooting medium carbon\n",
       " 72             rooting medium regulators\n",
       " 73             rooting medium solidifier\n",
       " 74                     rooting medium pH\n",
       " 75               air temperature regimen\n",
       " 84              mineral nutrient regimen\n",
       " 94                       plant treatment\n",
       " 95                         light regimen\n",
       " Name: Field Name, dtype: object,\n",
       " 2     NaN\n",
       " 4     NaN\n",
       " 7     NaN\n",
       " 12    NaN\n",
       " 13    NaN\n",
       " 14    NaN\n",
       " 15    NaN\n",
       " 16    NaN\n",
       " 19    NaN\n",
       " 22    NaN\n",
       " 23    NaN\n",
       " 25    NaN\n",
       " 26    NaN\n",
       " 29    NaN\n",
       " 34    NaN\n",
       " 35    NaN\n",
       " 37    NaN\n",
       " 39    NaN\n",
       " 41    NaN\n",
       " 45    NaN\n",
       " 46    NaN\n",
       " 47    NaN\n",
       " 48    NaN\n",
       " 49    NaN\n",
       " 50    NaN\n",
       " 51    NaN\n",
       " 52    NaN\n",
       " 53    NaN\n",
       " 54    NaN\n",
       " 56    NaN\n",
       " 57    NaN\n",
       " 58    NaN\n",
       " 59    NaN\n",
       " 60    NaN\n",
       " 61    NaN\n",
       " 62    NaN\n",
       " 63    NaN\n",
       " 76    NaN\n",
       " 77    NaN\n",
       " 78    NaN\n",
       " 79    NaN\n",
       " 80    NaN\n",
       " 81    NaN\n",
       " 82    NaN\n",
       " 83    NaN\n",
       " 85    NaN\n",
       " 86    NaN\n",
       " 87    NaN\n",
       " 88    NaN\n",
       " 89    NaN\n",
       " 90    NaN\n",
       " 91    NaN\n",
       " 93    NaN\n",
       " 96    NaN\n",
       " 97    NaN\n",
       " 98    NaN\n",
       " 99    NaN\n",
       " Name: id, dtype: object,\n",
       " 2                      extrachromosomal elements\n",
       " 4     sample volume or weight for DNA extraction\n",
       " 7                 geographic location (altitude)\n",
       " 12                                 identified_by\n",
       " 13                   geographic location (depth)\n",
       " 14                           environment (biome)\n",
       " 15                         environment (feature)\n",
       " 16               geographic location (elevation)\n",
       " 19                    sample material processing\n",
       " 22            amount or size of sample collected\n",
       " 23                       sample storage duration\n",
       " 25                       sample storage location\n",
       " 26                           sampling time point\n",
       " 29                                   sampled age\n",
       " 34                               sample wet mass\n",
       " 35                               sample dry mass\n",
       " 37                                 sample length\n",
       " 39                         sample capture status\n",
       " 41                          genetic modification\n",
       " 45                             biological status\n",
       " 46                            organism phenotype\n",
       " 47                                ancestral data\n",
       " 48                   source material description\n",
       " 49                           biotic relationship\n",
       " 50                                  growth habit\n",
       " 51                                     plant sex\n",
       " 52                           climate environment\n",
       " 53                           gaseous environment\n",
       " 54                          seasonal environment\n",
       " 56           soil_taxonomic/local classification\n",
       " 57    soil_taxonomic/local classification method\n",
       " 58                                     soil type\n",
       " 59                              soil type method\n",
       " 60                       drainage classification\n",
       " 61                                       texture\n",
       " 62                                texture method\n",
       " 63                            soil water content\n",
       " 76                            antibiotic regimen\n",
       " 77                              chemical mutagen\n",
       " 78                            fertilizer regimen\n",
       " 79                             fungicide regimen\n",
       " 80                                       gravity\n",
       " 81                        growth hormone regimen\n",
       " 82                             herbicide regimen\n",
       " 83                              humidity regimen\n",
       " 85                  non-mineral nutrient regimen\n",
       " 86                             pesticide regimen\n",
       " 87                                    pH regimen\n",
       " 88                             radiation regimen\n",
       " 89                              rainfall regimen\n",
       " 90                                  salt regimen\n",
       " 91                        standing water regimen\n",
       " 93                     water temperature regimen\n",
       " 96                                biotic regimen\n",
       " 97                             mechanical damage\n",
       " 98                       chemical administration\n",
       " 99                                  perturbation\n",
       " Name: Field Name, dtype: object]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isin = mapping[\"term_accession\"].isin(json_table[json_table[\"in_ena\"]!=False].groupby(['term_accession']).count().index.to_series())\n",
    "[mapping[np.invert(isin)][mapping[\"Requirement\"]==\"mandatory\"]['id'],\n",
    "mapping[np.invert(isin)][mapping[\"Requirement\"]==\"mandatory\"]['Field Name'],\n",
    "mapping[np.invert(isin)][mapping[\"Requirement\"]==\"recommended\"]['id'],\n",
    "mapping[np.invert(isin)][mapping[\"Requirement\"]==\"recommended\"]['Field Name'],\n",
    "mapping[np.invert(isin)][mapping[\"Requirement\"]==\"optional\"]['id'],\n",
    "mapping[np.invert(isin)][mapping[\"Requirement\"]==\"optional\"]['Field Name'],\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "debbf7e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 'NFDI4PSO:0000076',\n",
       " 'NFDI4PSO:0000075',\n",
       " nan,\n",
       " 'NFDI4PSO:0000074',\n",
       " 'NFDI4PSO:0002003',\n",
       " 'NFDI4PSO:0002004',\n",
       " 'NFDI4PSO:0000074',\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 'NFDI4PSO:0000009',\n",
       " nan,\n",
       " 'NFDI4PSO:0000065',\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 'NFDI4PSO:0000011',\n",
       " nan,\n",
       " nan,\n",
       " 'PO:0009011',\n",
       " 'PO:0009012',\n",
       " nan,\n",
       " 'NFDI4PSO:0000073',\n",
       " 'NFDI4PSO:0000071',\n",
       " nan,\n",
       " 'NFDI4PSO:0000072',\n",
       " nan,\n",
       " nan,\n",
       " 'TO:1000012',\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 'NFDI4PSO:0000031',\n",
       " nan,\n",
       " 'NFDI4PSO:0000030',\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 'PECO:0007147',\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " 'PECO:0007383',\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan,\n",
       " nan]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(mapping[\"id\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
